File: gfg/ConstructaBinarySearchTreefromgivenpostorder.cpp
#include "../lib/createTree.hpp"
#include "../lib/createVector.hpp"
using namespace std;

class Solution {
public:
  TreeNode *constructTree(vector<int> &arr) {
    TreeNode *root = constructTreeHelper(arr, INT_MIN, INT_MAX);
    return root;
  }

  TreeNode *constructTreeHelper(vector<int> &arr, int left, int right) {
    if (arr.empty()) {
      return nullptr;
    }

    int last = arr.back();
    if (last < left || last > right) {
      return nullptr;
    }

    TreeNode *node = new TreeNode(last);
    arr.pop_back();
    node->right = constructTreeHelper(arr, last, right);
    node->left = constructTreeHelper(arr, left, last);
    return node;
  }
};

int main(int argc, char **argv) {
  string line;
  while (getline(cin, line)) {
    vector<int> arr = stringToIntegerVector(line);
    TreeNode *root = Solution().constructTree(arr);
    cout << treeNodeToString(root) << endl;
  }
}

File: gfg/ConstructaBinarySearchTreeFromGivenPostorder.java
import java.io.BufferedInputStream;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class ConstructaBinarySearchTreeFromGivenPostorder {
  // * node class
  static class Node {
    int data;
    Node left, right;
    Node(int d) {
      data = d;
      left = right = null;
    }
  }

  static class Solution {
    private int n = 0;

    Solution(int n) {
      this.n = n;
    }

    public Node constructTree(int post[], int n) {
      Node root = constructTreeHelper(post, Integer.MIN_VALUE, Integer.MAX_VALUE);
      return root;
    }

    public Node constructTreeHelper(int post[], int start, int end) {
      if ( n ==  -1 || post[n] < start || post[n] > end) {
        return null;
      }

      int val = post[n];
      Node node = new Node(post[n]);
      n = n - 1;

      node.right = constructTreeHelper(post, val, end);
      node.left = constructTreeHelper(post, start, val);

      return node;
    }
  }

  public static void printTree(Node node) {
    if (node == null) {
      return;
    }

    System.out.print(node.data + ": ");
    if (node.left != null) {
      System.out.print(node.left.data + ", ");
    } else {
      System.out.print("null, ");
    }

    if (node.right != null) {
      System.out.println(node.right.data + ", ");
    } else {
      System.out.println("null, ");
    }

    printTree(node.left);
    printTree(node.right);
  }

  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

    String line;
    while ((line = br.readLine())  != null) {
      int[] arr = Arrays.stream(line.split(" ")).mapToInt(Integer::parseInt).toArray();

      Node root = new Solution(arr.length - 1).constructTree(arr, arr.length - 1);
      printTree(root);
    }
  }
}